### PR TITLE
[INDEXING-117] persist raw data to datastore

### PR DESCRIPTION
**Pull request by deelawn**
Commit history might look weird; I had to go back and rebase it off the target branch of this PR so all the changes in the diff are only my changes.

### CODE REVIEW

File: flow/chain.go [hunk 1]
```diff
@@ -6,25 +6,24 @@ import (
 
 	"github.com/figment-networks/polkadot-worker/api/mapper"
 	"github.com/figment-networks/polkadot-worker/api/scale"
+	"github.com/figment-networks/polkadot-worker/structs"
 )
 
-func (f *PolkadotFlow) GetChainMeta(ctx context.Context, height uint64) (pcm scale.PolkaChainMeta, err error) {
+func (f *PolkadotFlow) GetChainMeta(
+	ctx context.Context,
+	heightCtx structs.HeightFetcherContext,
```
Comment (lukanus) [inline flow/chain.go:14]: cosmetics: 
Naming this "Context" might be confusing

File: flow/chain.go [hunk 2]
```diff
@@ -6,25 +6,24 @@ import (
...
-	ddr, prm, err := f.client.GetOthers(ctx, sess, hi.BlockHash, hi.ParentHash, hi.GrandparentHash)
-	if err != nil {
-		return pcm, err
-	}
-	meta, err := f.client.GetMetadata(ctx, sess, hi.BlockHash, prm.SpecName, uint(prm.SpecVersion))
-	if err != nil {
-		return pcm, fmt.Errorf("error while getting metadata: %w", err)
-	}
-	return mapper.MapChainMeta(ddr, int(prm.SpecVersion), &meta.Decoder.Metadata)
+	pcm, err = mapper.MapChainMeta(*heightCtx.DDR, int(prv.SpecVersion), &meta.Decoder.Metadata)
```
Comment (lukanus) [inline flow/chain.go:27]: if heightCtx.DDR != nil error

File: flow/fetch_flow.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	"go.uber.org/zap"
+)
+
+type FetchFlow struct {
+	client Client
+	fe     Fetcher
+
+	ChainID         string
+	Network         string
+	datastoreClient datastore.DatastoreService_StoreRecordsClient
```
Comment (lukanus) [inline flow/fetch_flow.go:18]: Underscore ? :P

File: go.mod [hunk 4]
```diff
@@ -22,3 +22,8 @@ require (
 	google.golang.org/grpc v1.42.0
 	google.golang.org/protobuf v1.27.1
 )
+
+replace (
```
Comment (lukanus) [inline go.mod:26]: probably you'd like to remove those ultimately :)

File: api/api.go [hunk 5]
```diff
@@ -431,6 +431,11 @@ func (c *CSession) RouteResponse(r Response) (err error) {
 	return nil
 }
 
+// Limit is used so the sender knows how many requests can be sent.
+func (c *CSession) Limit() int {
```
Comment (lukanus) [inline api/api.go:435]: or just change limit to Limit :)

File: flow/fetch_flow.go (hunk 3 above)
Comment (deelawn) [reply flow/fetch_flow.go:18]: 🤷‍♂️  that's protobuf generated code

Comment (lukanus) [reply flow/fetch_flow.go:18]: Ah! :)

File: flow/flow.go [hunk 6]
```diff
@@ -155,97 +167,54 @@ func (f *PolkadotFlow) BlockAndTx(ctx context.Context, height uint64) (block str
...
-		errString := "[SUBSTRATE-FLOW] Error sending datastore message (GetRawData) "
-		f.logger.Error(errString, zap.Error(err), zap.Uint64("height", height))
-		return blockWM, fmt.Errorf("%s: %d %w", errString, height, err)
+		errString := "[SUBSTRATE-FLOW] Error retrieving data (GetRange)"
+		f.logger.Error(errString, zap.Error(err), zap.Uint64("height", hr.StartHeight))
+		err = fmt.Errorf("%s: %d %w", errString, hr.StartHeight, err)
+		return
 	}
 
 	acksResponse, err := client.CloseAndRecv()
```
[the remark this answers is not in the document]
Comment (deelawn) [reply flow/flow.go:203]: Yes, will change.

File: api/api.go (hunk 5 above)
Comment (deelawn) [reply api/api.go:435]: `CSession` is being used to implement some interface, so an exported `Limit` field wouldn't be available.

File: go.mod (hunk 4 above)
Comment (deelawn) [reply go.mod:26]: Indeed

File: flow/getBlock.go [hunk 7]
```diff
@@ -10,63 +10,99 @@ import (
...
 
-		return nil, fmt.Errorf("error while decoding data: %w", err)
+func (f *PolkadotFlow) GetBlock(
+	ctx context.Context,
+	hfc pstructs.HeightFetcherContext,
+) (block *blockpb.GetByHeightResponse, err error) {
+
+	sess := f.sg.GetSession()
+	defer sess.End()
+
```
[the remark this answers is not in the document]
Comment (deelawn) [reply flow/getBlock.go:44]: 👍

File: flow/chain.go (hunk 1 above)
Comment (deelawn) [reply flow/chain.go]: Renamed

File: flow/height.go [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+import (
+	"context"
+
+	"github.com/figment-networks/polkadot-worker/api/scale"
+	"github.com/figment-networks/polkadot-worker/structs"
+)
+
+func (f *PolkadotFlow) GetHeightInfo(ctx context.Context, height uint64) (structs.HeightInfo, error) {
+
+	sess := f.sg.GetSession()
```
[the remark this answers is not in the document]
Comment (deelawn) [reply flow/height.go:14]: @lukanus I've pushed the changes that address this issue.

**Pull request merged**